applications/PFEM2Application/custom_conditions/autoslip_inlet_3d.h

```diff
@@ -204,15 +204,15 @@ class MonolithicAutoSlipInlet3D : public Condition
       */
     virtual void CalculateLocalSystem(MatrixType& rLeftHandSideMatrix,
                                       VectorType& rRightHandSideVector,
-                                      ProcessInfo& rCurrentProcessInfo) override;
```

Comment:
I just stumbled upon this pull request and don't exactly know the logic behind the code, but watch out with this fix.

The function it's supposed to overwrite is ```Condition::CalculateLocalSystem(MatrixType&,VectorType&,const ProcessInfo&)```. Notice that the difference is the ```const``` qualifier of ```ProcessInfo```. While deleting the ```override``` does allow it to be compiled, this function won't be called if someone passes a ```const ProcessInfo&``` to it and the base class' ```CalculateLocalSystem``` will be executed.

Reply:
Hello @matekelemen,
again thanks for your input. I misinterpreted the error in the compilation. I understand the role of the override instruction now. However, I'd like to comment on something that you've said. You've said that if I delete the override, it will compile and the base class' CalculateLocalSystem will be called. However, in this case, the base class Element, has the function declared as: 
virtual void CalculateLocalSystem(MatrixType& rLeftHandSideMatrix,
                                      VectorType& rRightHandSideVector,
                                      const ProcessInfo& rCurrentProcessInfo)
                            
I really don't understand how come the code compiled in the first place. Shouldn't the compiler have given me an error? Because the version that I had with override deleted, didn't exist anywhere in the code. I'd be glad to hear your comment.
Regards,
Deniz

Reply:
Hi @dcagri,

Let's say you have a base class ```Base``` with a member function ```f```. You plan to inherit from ```Base``` and want ```f``` to behave differently for each derived class, so you declare ```f``` as ```virtual```:
```
class Base
{
public:
    virtual void f( const int& rX );
};
```
As mentioned, you have a class ```Derived``` that inherits from ```Base```, and specializes ```f```:
```
class Derived : public Base
{
public:
    virtual void f( const int& rX ) override;
};
```

Now if you have an instance of ```Derived``` and a ```Base*``` pointer to it, ```Derived::f``` will be called as expected. (Note that ```override``` doesn't really do anything that would affect the behaviour of the code, you can get rid of it and the program would run just like before. It's just there to tell the compiler that the function is meant to override a function from the base class with identical signature.)
```
Derived derived;
Base* p_derived = &derived;
int x;
p_derived->f(x);    // Calls Derived::f because an implicit const cast if performed on x
```

The complications begin when the ```const``` qualifier is left off the argument, because the two function's signature would no longer be identical:
```
class Base
{
public:
    virtual void f( const int& rX );
};

class Derived : public Base
{
public:
    virtual void f( int& rX ) override;
};
```

This won't compile as ```Derived::f``` doesn't override anything (no function with identical signature in the base class). If you remove the ```override``` specifier, the code becomes completely valid but semantically incorrect: you just created an overload of ```f``` that takes an ```int&``` instead of a ```const int&```.

The scary part starts here: the code compiles just fine but it won't do what you expect it to. Same as before, if you have a ```Base*``` pointer to an instance of ```Derived``` and try to call ```f```, the compiler will start looking for valid declarations of ```f``` from the root of the inheritance hierarchy (i.e.: starts looking for a suitable ```f``` in ```Base``` first). It finds ```Base::f``` and checks its [vtable](https://stackoverflow.com/a/3555290) but doesn't find an override (since ```Derived::f``` doesn't have the same signature), so it does a ```const``` cast on the argument and calls ```Base::f```. [example](https://godbolt.org/z/hKG7jv).
```
Derived derived;
Base* p_derived = &derived;
int x;
p_derived->f(x);    // Calls Base::f
```

The only way you can call ```Derived::f``` is if you have access to the instance itself or hold a ```Derived*``` pointer to it. Furthermore, you can only call it with a non-const ```int&``` and will get a compiler error if you try ```const int&``` because of shadowing, but that's a different story.
```
Derived derived;
Base* p_derived = &derived;
Derived* p_trueDerived = &derived;
int x = 0;
const int cx = 0;

p_derived->f(x);    // Calls Base::f
p_derived->f(cx);    // Calls Base::f
p_trueDerived->f(x);    // Calls Derived::f
p_trueDerived->f(cx);    // Error: no matching function to call for Derived::f(const int&)
```

Not sure whether I answered your question but I hope you find this useful.

Reply:
Hi @matekelemen,
yes, you did answer my question and it was very helpful!!!
Thank you very much and have a nice evening.
Regards,
Deniz